### PR TITLE
Gudermannian tweaks

### PR DESCRIPTION
**Pull request by rocky**

### CODE REVIEW

**Comment by rocky** [comment]
I have been looking to figure out what it would take to get  Mathics3 align with all of the rules in https://github.com/axkr/symja_android_library/tree/master/symja_android_library/rules

In that, it looks the Complex constructor doesn't take Rational real and imaginary parts, so that has been added. 

A test for starting to test Gudermannian and more generally hyperbolic fns was started. 

There is a lot more, and many unsolved problems. But I wanted to jot down what I have been able to detect so far.

File: mathics/builtin/specialfns/bessel.py [hunk 1]
```diff
@@ -341,14 +341,14 @@ class BesselI(_Bessel):
...
-    rules = {
-        "Derivative[0, 1][BesselI]": "((BesselI[-1 + #1, #2] + BesselI[1 + #1, #2])/2)&",
-    }
-
     mpmath_name = "besseli"
     rules = {
         "BesselI[Undefined, x_]": "Undefined",
         "BesselI[y_, Undefined]": "Undefined",
+        # FIXME: thes are not respected. Why?
+        "BesselI[x_, -I Infinity]": "0",
```
Comment (mmatera) [inline mathics/builtin/specialfns/bessel.py:349]: What about using `DirectedInfinity[I]`?

File: mathics/core/atoms.py [hunk 2]
```diff
@@ -672,12 +672,12 @@ class Complex(Number):
...
     def __new__(cls, real, imag):
-        if isinstance(real, Complex) or not isinstance(real, Number):
-            raise ValueError("Argument 'real' must be a Real number.")
+        if not isinstance(real, Number):
+            raise ValueError(f"Argument 'real' must be a Number; is {real}.")
         if imag is SymbolInfinity:
             return SymbolI * SymbolInfinity
-        if isinstance(imag, Complex) or not isinstance(imag, Number):
-            raise ValueError("Argument 'imag' must be a Real number.")
+        if not isinstance(imag, Number):
```
Comment (mmatera) [inline mathics/core/atoms.py:679]: What about if imag is a complex number?

File: mathics/builtin/specialfns/bessel.py (hunk 1 above)
Comment (rocky) [reply mathics/builtin/specialfns/bessel.py:349]: Doesn't change things. Ditto for `DirectedInfiity[-I]`  and `-DirectedInfity[i]` and `-ComplexInfiity`. 

I wonder if Sympy handling is skipping rule application. I see  in other situations too.

File: mathics/core/atoms.py (hunk 2 above)
Comment (rocky) [reply mathics/core/atoms.py:679]: Good point! I will change this to `(Integer, Real, Rational)`

File: mathics/builtin/specialfns/bessel.py (hunk 1 above)
Comment (mmatera) [reply mathics/builtin/specialfns/bessel.py:349]: What does TraceEvaluation show?

Comment (rocky) [reply mathics/builtin/specialfns/bessel.py:349]:
```
In[2]:= TraceEvaluation[BesselI[x, -I Infinity]]
    Evaluating: System`BesselI[Global`x, System`Times[-1, System`I, System`Infinity]]
      Evaluating: System`BesselI
      Evaluating: Global`x
      Evaluating: System`Times[-1, System`I, System`Infinity]
        Evaluating: System`Times
        Evaluating: System`I
        -> I
        Evaluating: System`Infinity
        -> System`DirectedInfinity[1]
        Evaluating: System`DirectedInfinity[1]
          Evaluating: System`DirectedInfinity
          Evaluating: System`Unequal[System`N[System`Abs[1]], 1]
            Evaluating: System`Unequal
            Evaluating: System`N[System`Abs[1]]
              Evaluating: System`N
              Evaluating: System`Abs[1]
                Evaluating: System`Abs
            Evaluating: System`Function[System`Not[System`ExactNumberQ[System`Slot[1]]]][1.0]
              Evaluating: System`Function[System`Not[System`ExactNumberQ[System`Slot[1]]]]
                Evaluating: System`Function
            -> System`Not[System`ExactNumberQ[1.0]]
              Evaluating: System`Not
              Evaluating: System`ExactNumberQ[1.0]
                Evaluating: System`ExactNumberQ
              -> System`False
            -> System`True
            Evaluating: System`Function[System`Not[System`ExactNumberQ[System`Slot[1]]]][1]
              Evaluating: System`Function[System`Not[System`ExactNumberQ[System`Slot[1]]]]
                Evaluating: System`Function
            -> System`Not[System`ExactNumberQ[1]]
              Evaluating: System`Not
              Evaluating: System`ExactNumberQ[1]
                Evaluating: System`ExactNumberQ
              -> System`True
            -> System`False
            Evaluating: System`Function[System`Not[System`ExactNumberQ[System`Slot[1]]]][1.0]
              Evaluating: System`Function[System`Not[System`ExactNumberQ[System`Slot[1]]]]
                Evaluating: System`Function
            -> System`Not[System`ExactNumberQ[1.0]]
              Evaluating: System`Not
              Evaluating: System`ExactNumberQ[1.0]
                Evaluating: System`ExactNumberQ
              -> System`False
            -> System`True
            Evaluating: System`ExactNumberQ[1.0]
              Evaluating: System`ExactNumberQ
            -> System`False
            Evaluating: System`ExactNumberQ[1]
              Evaluating: System`ExactNumberQ
            -> System`True
            Evaluating: System`$MaxExtraPrecision
          -> System`False
      -> System`DirectedInfinity[-I]
        Evaluating: System`DirectedInfinity
        Evaluating: System`Unequal[System`N[System`Abs[-I]], 1]
          Evaluating: System`Unequal
          Evaluating: System`N[System`Abs[-I]]
            Evaluating: System`N
            Evaluating: System`Abs[-I]
              Evaluating: System`Abs
          Evaluating: System`Function[System`Not[System`ExactNumberQ[System`Slot[1]]]][1.0]
            Evaluating: System`Function[System`Not[System`ExactNumberQ[System`Slot[1]]]]
              Evaluating: System`Function
          -> System`Not[System`ExactNumberQ[1.0]]
            Evaluating: System`Not
            Evaluating: System`ExactNumberQ[1.0]
              Evaluating: System`ExactNumberQ
            -> System`False
          -> System`True
          Evaluating: System`Function[System`Not[System`ExactNumberQ[System`Slot[1]]]][1]
            Evaluating: System`Function[System`Not[System`ExactNumberQ[System`Slot[1]]]]
              Evaluating: System`Function
          -> System`Not[System`ExactNumberQ[1]]
            Evaluating: System`Not
            Evaluating: System`ExactNumberQ[1]
              Evaluating: System`ExactNumberQ
            -> System`True
          -> System`False
          Evaluating: System`Function[System`Not[System`ExactNumberQ[System`Slot[1]]]][1.0]
            Evaluating: System`Function[System`Not[System`ExactNumberQ[System`Slot[1]]]]
              Evaluating: System`Function
          -> System`Not[System`ExactNumberQ[1.0]]
            Evaluating: System`Not
            Evaluating: System`ExactNumberQ[1.0]
              Evaluating: System`ExactNumberQ
            -> System`False
          -> System`True
          Evaluating: System`ExactNumberQ[1.0]
            Evaluating: System`ExactNumberQ
          -> System`False
          Evaluating: System`ExactNumberQ[1]
            Evaluating: System`ExactNumberQ
          -> System`True
          Evaluating: System`$MaxExtraPrecision
        -> System`False
      Evaluating: Global`x
      Evaluating: System`ComplexInfinity
      -> System`DirectedInfinity[]
      Evaluating: System`DirectedInfinity[]
        Evaluating: System`DirectedInfinity
    -> System`BesselI[Global`x, System`DirectedInfinity[]]
      Evaluating: System`BesselI
      Evaluating: Global`x
      Evaluating: System`DirectedInfinity[]
      Evaluating: Global`x
      Evaluating: System`ComplexInfinity
      -> System`DirectedInfinity[]
      Evaluating: System`DirectedInfinity[]
        Evaluating: System`DirectedInfinity
Out[2]= BesselI[x, ComplexInfinity]
```

Comment (mmatera) [reply mathics/builtin/specialfns/bessel.py:349]: Notice that most of the evaluation is about convert `I infinity` into `ComplexInfinity`, which is also wrong. That are the rules to be fixed.

Comment (rocky) [reply mathics/builtin/specialfns/bessel.py:349]: Ok. I will leave it for you or someone else (or maybe me in the future) to fix.

**Pull request merged**